### ISSUE TITLE
Q: performance difference forward vs playback from file

### ISSUE THREAD

**Issue by kjwierenga** [buger/goreplay#548]
**TL;DR** To get production comparable performance I need to set `-input-raw-expire` to a lower value (< 500ms). Why?

I've noticed a distinct difference in the performance of our (Rails) stack forwarding traffic from production to staging directly (`-input-raw` to `-output-http`) and capture and playback (`-input-raw` to `-output-file` then `-input-file` to `-output-http`).

I have the following load test running on one of our production nodes.

    ./goreplay -input-raw :80 -output-http="https://our-staging-server.com/"

The performance (as viewed/measured with NewRelic) shows 3 times slower average response time from our Rails app (~120ms/request in staging vs ~40ms/request in production). The production and staging setups are the same and should perform equally.

When I repeat the same test by capture to file and playback:

    ./goreplay -input-raw :80 -output-file requests.log
    ./goreplay -input-file requests_0.log -output-http="https://our-staging-server.com/"

The performance with this setup is the same in both environments (~40ms/request in both production and staging).

I've also discovered that reducing `-input-raw-expire` timeout to 200ms (from default 2s), the issue is resolved and the same performance is reported when forwarding live requests from production to staging.

    ./goreplay -input-raw-expire 0.2s -input-raw:80 -output-http="https://our-staging-server.com"

With these parameters we see normal (~40ms/request) response times reported in both production and staging.

What could be the cause for these differences? Is there a way for `goreplay` to yield the expected behaviour with a default setup?

**Comment by buger** [comment]
Hello! This is indeed weird since nature of this option is quite different. 

The basic idea here is that some requests take longer, for example, file upload or long polling and etc. Expiration, in this case, means how long GoReplay should wait for this request to complete. 

I would recommend looking at percentiles instead of averages, to see if some of the requests take longer on staging. With bigger --input-raw-expire it may catch more requests, especially slow ones. It also makes sense manually check recorded files, and try to detect such requests by looking at latency.

However, I'm not sure why replaying from file helps you. Response latency is a server side thing, GoRelpay just emits requests. 

Also worth considering theory that while replaying from live, it replay on a higher rate than from file (in theory it should be same, but maybe there is bug somewhere), and when you set `-input-raw-expire` to lower value it drops some requests. Try also to measure requests per second coming to staging in both cases, maybe it has bigger latency just because can't handle requests on the same rate as production? 

Hope it helps!

**Comment by Lexing** [comment]
I've encountered the exactly same problem

Using -input-raw to replay traffic is about 2x~3x slower than production (same set up on real traffic without goreplay)

I expect the performance of two to be the same though. What's special in goreply's connection ?

**Comment by avnerbarr** [comment]
we saw this as well. we forwarded traffic from one production server to another production server which was disconnected from the cluster incoming load balancer. The response times for the test case went up ~5x

**Comment by itayB** [comment]
I've added this flag and the response time changed dramatically: --output-http-workers=2
more info: https://github.com/buger/goreplay/wiki/Replaying-HTTP-traffic
